src/tests/mission.cpp

```diff
@@ -6,49 +6,91 @@
 #include <vector>
 
 using namespace std;
-using namespace dronecode_sdk;
 
 namespace tests
 {
-REGISTER_TEST(MissionUpload);
+REGISTER_TEST(Mission);
 
-MissionUpload::MissionUpload(const Context& context) : TestBase(context), _mission(context.system) {}
+Mission::Mission(const Context& context) : TestBase(context), _mission(context.system) {}
 
-shared_ptr<MissionItem> MissionUpload::makeMissionItem(double latitude_deg, double longitude_deg,
-                                                       float relative_altitude_m)
+shared_ptr<dcsdk::MissionItem> Mission::makeMissionItem(double latitude_deg, double longitude_deg,
+                                                        float relative_altitude_m)
 {
-    shared_ptr<MissionItem> new_item = make_shared<MissionItem>();
+    shared_ptr<dcsdk::MissionItem> new_item = make_shared<dcsdk::MissionItem>();
     new_item->set_position(latitude_deg, longitude_deg);
     new_item->set_relative_altitude(relative_altitude_m);
     return new_item;
 }
 
-TestBase::Result MissionUpload::run()
+void Mission::run()
+{
+    std::vector<std::shared_ptr<dcsdk::MissionItem>> items = assembleMissionItems();
+
+    uploadMission(items);
+    std::vector<std::shared_ptr<dcsdk::MissionItem>> downloaded_items = downloadMission();
+
+    compareMissions(items, downloaded_items);
+}
+
+std::vector<std::shared_ptr<dcsdk::MissionItem>> Mission::assembleMissionItems()
 {
     cout << "Number of waypoints: " << _config.num_waypoints << endl;
 
-    vector<shared_ptr<MissionItem>> mission_items;
+    std::vector<std::shared_ptr<dcsdk::MissionItem>> items{};
 
     for (int i = 0; i < _config.num_waypoints; ++i) {
         float altitude = 10.f + (float)i;
         double latitude = 47.398170327054473 + (double)i * 1e-5;
-        mission_items.push_back(makeMissionItem(latitude, 8.5456490218639658, altitude));
+        items.push_back(makeMissionItem(latitude, 8.5456490218639658, altitude));
     }
 
+    return items;
+}
+
+void Mission::uploadMission(const std::vector<std::shared_ptr<dcsdk::MissionItem>>& items)
+{
     cout << "Uploading mission..." << endl;
-    auto prom = make_shared<promise<Mission::Result>>();
-    auto future_result = prom->get_future();
-    _mission.upload_mission_async(mission_items, [prom](Mission::Result result) { prom->set_value(result); });
+    promise<dcsdk::Mission::Result> prom{};
+    auto fut = prom.get_future();
+
+    _mission.upload_mission_async(
+        items, [&prom](dcsdk::Mission::Result result) { prom.set_value(result); });
 
     // wait until uploaded
-    const Mission::Result result = future_result.get();
-    if (result != Mission::Result::SUCCESS) {
-        cout << "Mission upload failed (" << Mission::result_str(result) << "), exiting." << endl;
-        return Result::Failed;
-    }
-    cout << "Mission uploaded." << endl;
+    const dcsdk::Mission::Result result = fut.get();
+    EXPECT_EQ(result, dcsdk::Mission::Result::SUCCESS);
```

Comment:
If the upload fails, there's probably no point in downloading it. Meaning this should be an assertion.

Reply:
It might have been uploaded anyway but we got a timeout. I think we can still try to download it, and most probably it'll be wrong.